core/src/subgraph/context/instance.rs

```
@@ -204,11 +202,121 @@ where
+/// order.
+///
+/// This structure also maintains a partition of the hosts by address, for faster trigger matching.
+/// This partition uses the host's index in the main vec, to maintain the correct ordering.
+struct Hosts<C: Blockchain, T: RuntimeHostBuilder<C>> {
+    hosts: Vec<Arc<T::Host>>,
+
+    // The `usize` is the index of the host in `hosts`.
+    hosts_by_address: HashMap<Box<[u8]>, Vec<usize>>,
+    hosts_without_address: Vec<usize>,
+}
+
+impl<C: Blockchain, T: RuntimeHostBuilder<C>> Hosts<C, T> {
+    fn new() -> Self {
+        Self {
+            hosts: Vec::new(),
+            hosts_by_address: HashMap::new(),
+            hosts_without_address: Vec::new(),
+        }
+    }
+
+    fn hosts(&self) -> &[Arc<T::Host>] {
+        &self.hosts
+    }
+
+    fn last(&self) -> Option<&Arc<T::Host>> {
+        self.hosts.last()
+    }
+
+    fn len(&self) -> usize {
+        self.hosts.len()
+    }
+
+    fn push(&mut self, host: Arc<T::Host>) {
+        self.hosts.push(host.cheap_clone());
+        let idx = self.hosts.len() - 1;
+        let address = host.data_source().address();
+        match address {
+            Some(address) => {
+                self.hosts_by_address
+                    .entry(address.into())
+                    .or_default()
+                    .push(idx);
+            }
+            None => {
+                self.hosts_without_address.push(idx);
+            }
+        }
+    }
+
+    fn pop(&mut self) {
+        let Some(host) = self.hosts.pop() else { return };
+        let address = host.data_source().address();
+        match address {
+            Some(address) => {
+                // Unwrap and assert: The same host we just popped must be the last one in `hosts_by_address`.
+                let hosts = self.hosts_by_address.get_mut(address.as_slice()).unwrap();
+                let idx = hosts.pop().unwrap();
+                assert_eq!(idx, self.hosts.len());
```

Is this running inside some panic handler that will stop the subgraph if this assert gets tripped? Otherwise, it would be better to return an error so that only this subgraph stops, and others can continue

---

Thanks for the review, I've checked and this panic will only stop the runner thread for the subgraph that trips it.